### ISSUE TITLE
The admin site randomly stops working: AdminMenuFilter causing Microsoft.CSharp.RuntimeBinder.RuntimeBinderException?

### ISSUE THREAD

**Issue by mikoskinen** [OrchardCMS/OrchardCore#5016]
We're running Orchard Core CMS on Azure App Service and we've now encountered the same issue few times: The admin-side of the site stops randomly working. The site itself still continues to work OK, but when we try to access /admin, we get an exception. After restarting the App Service, everything again works.

This is what we can see in the logs:

```
2019-12-11 12:42:05.2677|1|ERROR|Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware|An unhandled exception has occurred while executing the request. Microsoft.CSharp.RuntimeBinder.RuntimeBinderException: Cannot perform runtime binding on a null reference
   at CallSite.Target(Closure , CallSite , Object , IShape )
   at OrchardCore.Admin.AdminMenuFilter.OnResultExecutionAsync(ResultExecutingContext filterContext, ResultExecutionDelegate next) in C:\projects\orchardcore\src\OrchardCore.Modules\OrchardCore.Admin\AdminMenuFilter.cs:line 78
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|29_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at OrchardCore.Diagnostics.DiagnosticsStartupFilter.<>c__DisplayClass3_0.<<Configure>b__1>d.MoveNext() in C:\projects\orchardcore\src\OrchardCore.Modules\OrchardCore.Diagnostics\DiagnosticsStartupFilter.cs:line 34
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)|url: https://mytestsite.azurewebsites.net/admin|action: Index
```

The stacktrace seems to point into these lines:

            dynamic layout = await _layoutAccessor.GetLayoutAsync();

            if (layout.Navigation is ZoneOnDemand zoneOnDemand)
            {
                await zoneOnDemand.AddAsync(menuShape);
            }
            else
            {
                layout.Navigation.Add(menuShape);
            }

And based on the exception, layout in our case seems to be null, causing layout.Navigation.Add to crash.

Any ideas why _layoutAccessor.GetLayoutAsync could randomly start returning null?

**Comment by Skrypt** [comment]
I've had this issue in the past and I ended up disabling the Layout module and use templates to integrate my zones directly. This is a known issue.

**Comment by deanmarcussen** [comment]
is it possible the `_layoutAccessor.GetLayoutAsync()` needs a lock to prevent reentry when it's null and still getting the layout? 

/cc @jtkech

**Comment by sebastienros** [comment]
Not the first time this issue is reported ... taking another quick look based on @deanmarcussen 's suggestion

**Comment by deanmarcussen** [comment]
this one isn't related to the layers module, but @Skrypt and @johnrom have also reported similar issues on https://github.com/OrchardCMS/OrchardCore/issues/3790 but relating to the layers module.

Wonder if it's related, however as it all seems to be about getting the layout accessor.

**Comment by sebastienros** [comment]
@mikoskinen what version are you using?

**Comment by mikoskinen** [comment]
@sebastienros 1.0.0-rc1-11022.

**Comment by sebastienros** [comment]
I found nothing that could explain this. Could you add some null checks to narrow down what is happening. Like on layout.Navigation, and also why it goes on line 78 of the filter.

I created a PR based on my review here: https://github.com/OrchardCMS/OrchardCore/pull/5022
It would throw if the layou is null, but I don't it's this. I think it's Layout.Navigation. This property should be created on demand though, with a type of `Zone`. You can also try my branch and see if it fixes it.

**Comment by jtkech** [comment]
Yes, here the layout is not null otherwise it would have 1st failed line 72. So, this would mean that here layout is not null but not a `ZoneHolding`.

> is it possible the _layoutAccessor.GetLayoutAsync() needs a lock to prevent reentry when it's null and still getting the layout?

I don't think so, it is a scoped service but maybe a concurrency issue in `.GetShapeTable()`.

By adding this in `DefaultShapeFactory.CreateAsync()`

            if (shapeType == "Layout")
            {
                shapeDescriptor = null;
            }

Indeed in this case layout is not null, it is a shape but not a `ZoneHolding`, **and i could repro the exact same issue**.

In `DefaultShapeTableManager.GetShapeTable()` we build a shape table once per theme and cache it in the shared memory cache. To not do duplicate works for another theme, we exclude the features whose shape descriptors have already been built. But if a thread is building the shape descriptors of a given feature (e.g the admin theme), another concurrent thread may exclude this feature while the 1st thread has not built all descriptors yet. Then, if this concurrent thread is the last one to set the memory cache, a descriptor may be missing.

Note: Most of the descriptors of all enabled features are built on the 1st request, only the descriptors related to themes may be built afterwards on request, so there is a "better" chance to have a concurrency issue with theme descriptors.

So we would need some locking, at least at the end checking again if the cache has not been set. Maybe use a tenant level lock but a named one (as implemented in #4863) based on the theme name.
